x-pack/plugins/fleet/server/services/epm/packages/integrations_state_machine.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { appContextService } from '../../app_context';
+
+export interface State {
+  event: any;
+  nextState?: string;
+  currentStatus?: string;
+}
+export type StatusName = 'success' | 'failed' | 'pending';
+
+export const installStateNames = [
+  'create_restart_installation',
+  'install_kibana_assets',
+  'install_ml_model',
+  'install_index_template_pipelines',
+  'remove_legacy_templates',
+  'update_current_write_indices',
+  'install_transforms',
+  'delete_previous_pipelines',
+  'save_archive_entries_from_assets_map',
+  'update_so',
+] as const;
+
+type StateNamesTuple = typeof installStateNames;
+type StateNames = StateNamesTuple[number];
+type StateMachineStates<T extends string> = Record<T, State>;
+
+// example our install states
+const installStates: StateMachineStates<StateNames> = {
```

Comment:
Should we expose a way to store any kind of context on the state machine that can be referenced in the event callbacks? For instance, maybe we could "hang" the `PackageInfo` object in this machine's context so it can be referenced by each state transition handler below.